File: wpilibc/src/test/native/cpp/simulation/GearboxSimTest.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include "wpi/simulation/GearboxSim.hpp"

#include <gtest/gtest.h>

#include "wpi/hardware/motor/PWMVictorSPX.hpp"
#include "wpi/hardware/rotation/Encoder.hpp"
#include "wpi/math/controller/PIDController.hpp"
#include "wpi/math/system/Gearbox.hpp"
#include "wpi/math/system/Models.hpp"
#include "wpi/simulation/BatterySim.hpp"
#include "wpi/simulation/EncoderSim.hpp"
#include "wpi/simulation/RoboRioSim.hpp"
#include "wpi/system/RobotController.hpp"

TEST(GearboxSimTest, VoltageSteadyState) {
  wpi::math::Gearbox gearbox{wpi::math::DCMotor::kNEO, 1, 1.0};
  auto plant = wpi::math::Models::SingleJointedArmFromPhysicalConstants(
      gearbox, wpi::units::kilogram_square_meter_t{0.0005});
  wpi::sim::GearboxSim sim{plant, gearbox};

  wpi::Encoder encoder{0, 1};
  wpi::sim::EncoderSim encoderSim{encoder};
  wpi::PWMVictorSPX motor{0};

  wpi::sim::RoboRioSim::ResetData();
  encoderSim.ResetData();

  // Spin-up
  for (int i = 0; i < 100; i++) {
    // RobotPeriodic runs first
    motor.SetVoltage(12_V);

    // Then, SimulationPeriodic runs
    wpi::sim::RoboRioSim::SetVInVoltage(
        wpi::sim::BatterySim::Calculate({sim.GetCurrentDraw()}));
    sim.SetInputVoltage(motor.GetThrottle() *
                        wpi::RobotController::GetBatteryVoltage());
    sim.Update(20_ms);
    encoderSim.SetRate(sim.GetAngularVelocity().value());
  }

  EXPECT_NEAR((gearbox.motor.Kv * 12_V).value(), encoder.GetRate(), 0.1);

  // Decay
  for (int i = 0; i < 100; i++) {
    // RobotPeriodic runs first
    motor.SetVoltage(0_V);

    // Then, SimulationPeriodic runs
    wpi::sim::RoboRioSim::SetVInVoltage(
        wpi::sim::BatterySim::Calculate({sim.GetCurrentDraw()}));
    sim.SetInputVoltage(motor.GetThrottle() *
                        wpi::RobotController::GetBatteryVoltage());
    sim.Update(20_ms);
    encoderSim.SetRate(sim.GetAngularVelocity().value());
  }

  EXPECT_NEAR(0, encoder.GetRate(), 0.1);
}

TEST(GearboxSimTest, PositionFeedbackControl) {
  wpi::math::Gearbox gearbox{wpi::math::DCMotor::kNEO, 1, 1.0};
  auto plant = wpi::math::Models::SingleJointedArmFromPhysicalConstants(
      gearbox, wpi::units::kilogram_square_meter_t{0.0005});
  wpi::sim::GearboxSim sim{plant, gearbox};

  wpi::math::PIDController controller{0.04, 0.0, 0.001};

  wpi::Encoder encoder{0, 1};
  wpi::sim::EncoderSim encoderSim{encoder};
  wpi::PWMVictorSPX motor{0};

  wpi::sim::RoboRioSim::ResetData();
  encoderSim.ResetData();

  for (int i = 0; i < 140; i++) {
    // RobotPeriodic runs first
    motor.SetThrottle(controller.Calculate(encoder.GetDistance(), 750));

    // Then, SimulationPeriodic runs
    wpi::sim::RoboRioSim::SetVInVoltage(
        wpi::sim::BatterySim::Calculate({sim.GetCurrentDraw()}));
    sim.SetInputVoltage(motor.GetThrottle() *
                        wpi::RobotController::GetBatteryVoltage());
    sim.Update(20_ms);
    encoderSim.SetDistance(sim.GetAngularPosition().value());
    encoderSim.SetRate(sim.GetAngularVelocity().value());
  }

  EXPECT_NEAR(encoder.GetDistance(), 750, 1.0);
  EXPECT_NEAR(encoder.GetRate(), 0, 0.1);
}

TEST(GearboxSimTest, CurrentDrawScalesWithMotorCount) {
  // Current draw is the total across the gearbox's motors, so doubling the
  // motor count while holding the mechanism identical should double the stall
  // current.
  constexpr auto J = wpi::units::kilogram_square_meter_t{0.0005};
  constexpr double kReduction = 3.0;

  wpi::math::Gearbox oneMotor{wpi::math::DCMotor::kNEO, 1, kReduction};
  wpi::math::Gearbox twoMotors{wpi::math::DCMotor::kNEO, 2, kReduction};

  wpi::sim::GearboxSim oneMotorSim{
      wpi::math::Models::SingleJointedArmFromPhysicalConstants(oneMotor, J),
      oneMotor};
  wpi::sim::GearboxSim twoMotorSim{
      wpi::math::Models::SingleJointedArmFromPhysicalConstants(twoMotors, J),
      twoMotors};

  // At rest, current draw is the stall current at the applied voltage.
  oneMotorSim.SetInput(wpi::math::Vectord<1>{12.0});
  twoMotorSim.SetInput(wpi::math::Vectord<1>{12.0});

  EXPECT_NEAR(wpi::math::DCMotor::kNEO.Current(0_rad_per_s, 12_V).value(),
              oneMotorSim.GetCurrentDraw().value(), 1e-9);
  EXPECT_NEAR(2.0 * oneMotorSim.GetCurrentDraw().value(),
              twoMotorSim.GetCurrentDraw().value(), 1e-9);
}

File: wpilibj/src/test/java/org/wpilib/simulation/GearboxSimTest.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package org.wpilib.simulation;

import static org.junit.jupiter.api.Assertions.assertEquals;

import org.junit.jupiter.api.Test;
import org.wpilib.hardware.motor.PWMVictorSPX;
import org.wpilib.hardware.rotation.Encoder;
import org.wpilib.math.controller.PIDController;
import org.wpilib.math.numbers.N1;
import org.wpilib.math.numbers.N2;
import org.wpilib.math.system.DCMotor;
import org.wpilib.math.system.Gearbox;
import org.wpilib.math.system.LinearSystem;
import org.wpilib.math.system.Models;
import org.wpilib.system.RobotController;

class GearboxSimTest {
  @Test
  void testVoltageSteadyState() {
    RoboRioSim.resetData();

    Gearbox gearbox = new Gearbox(DCMotor.kNEO, 1, 1);
    LinearSystem<N2, N1, N2> plant = Models.singleJointedArmFromPhysicalConstants(gearbox, 0.0005);
    GearboxSim sim = new GearboxSim(plant, gearbox);

    try (var motor = new PWMVictorSPX(0);
        var encoder = new Encoder(0, 1)) {
      var encoderSim = new EncoderSim(encoder);
      encoderSim.resetData();

      for (int i = 0; i < 100; i++) {
        motor.setVoltage(12);

        // ------ SimulationPeriodic() happens after user code -------
        RoboRioSim.setVInVoltage(
            BatterySim.calculateDefaultBatteryLoadedVoltage(sim.getCurrentDraw()));
        sim.setInputVoltage(motor.getThrottle() * RobotController.getBatteryVoltage());
        sim.update(0.020);
        encoderSim.setRate(sim.getAngularVelocity());
      }

      assertEquals(gearbox.motor.Kv * 12, encoder.getRate(), 0.1);

      for (int i = 0; i < 100; i++) {
        motor.setVoltage(0);

        // ------ SimulationPeriodic() happens after user code -------
        RoboRioSim.setVInVoltage(
            BatterySim.calculateDefaultBatteryLoadedVoltage(sim.getCurrentDraw()));
        sim.setInputVoltage(motor.getThrottle() * RobotController.getBatteryVoltage());
        sim.update(0.020);
        encoderSim.setRate(sim.getAngularVelocity());
      }

      assertEquals(0, encoder.getRate(), 0.1);
    }
  }

  @Test
  void testPositionFeedbackControl() {
    RoboRioSim.resetData();

    Gearbox gearbox = new Gearbox(DCMotor.kNEO, 1, 1);
    LinearSystem<N2, N1, N2> plant = Models.singleJointedArmFromPhysicalConstants(gearbox, 0.0005);
    GearboxSim sim = new GearboxSim(plant, gearbox);

    try (var motor = new PWMVictorSPX(0);
        var encoder = new Encoder(0, 1);
        var controller = new PIDController(0.04, 0.0, 0.001)) {
      var encoderSim = new EncoderSim(encoder);
      encoderSim.resetData();

      for (int i = 0; i < 140; i++) {
        motor.setThrottle(controller.calculate(encoder.getDistance(), 750));

        // ------ SimulationPeriodic() happens after user code -------
        RoboRioSim.setVInVoltage(
            BatterySim.calculateDefaultBatteryLoadedVoltage(sim.getCurrentDraw()));
        sim.setInputVoltage(motor.getThrottle() * RobotController.getBatteryVoltage());
        sim.update(0.020);
        encoderSim.setDistance(sim.getAngularPosition());
        encoderSim.setRate(sim.getAngularVelocity());
      }

      assertEquals(750, encoder.getDistance(), 1.0);
      assertEquals(0, encoder.getRate(), 0.1);
    }
  }

  @Test
  void testCurrentDrawScalesWithMotorCount() {
    // Current draw is the total across the gearbox's motors, so doubling the motor
    // count while holding the mechanism identical should double the stall current.
    double j = 0.0005;
    double reduction = 3.0;

    var oneMotor = new Gearbox(DCMotor.kNEO, 1, reduction);
    var twoMotors = new Gearbox(DCMotor.kNEO, 2, reduction);

    var oneMotorSim =
        new GearboxSim(Models.singleJointedArmFromPhysicalConstants(oneMotor, j), oneMotor);
    var twoMotorSim =
        new GearboxSim(Models.singleJointedArmFromPhysicalConstants(twoMotors, j), twoMotors);

    // At rest, current draw is the stall current at the applied voltage.
    oneMotorSim.setInput(12.0);
    twoMotorSim.setInput(12.0);

    assertEquals(DCMotor.kNEO.getCurrent(0.0, 12.0), oneMotorSim.getCurrentDraw(), 1e-9);
    assertEquals(2.0 * oneMotorSim.getCurrentDraw(), twoMotorSim.getCurrentDraw(), 1e-9);
  }
}
